infer/src/IR/Cfg.ml

```diff
@@ -30,11 +30,19 @@ let get_all_defined_proc_names cfg =
   !procs
 
 
+let signature_eq pdesc1 pdesc2 =
+  let old_attrs = Procdesc.get_attributes pdesc1 in
+  let new_attrs = Procdesc.get_attributes pdesc2 in
+  let formals_eq = List.equal (fun (_, typ, _) (_, typ', _) -> Typ.equal typ typ') in
+  (not (Typ.equal old_attrs.ret_type new_attrs.ret_type))
+  || not (formals_eq old_attrs.formals new_attrs.formals)
+
+
 let create_proc_desc cfg (proc_attributes : ProcAttributes.t) =
   let pdesc = Procdesc.from_proc_attributes proc_attributes in
   let pname = proc_attributes.proc_name in
-  if Procname.Hash.mem cfg pname then
-    L.die InternalError "Creating two procdescs for the same procname." ;
+  if Procname.Hash.mem cfg pname && signature_eq pdesc (Procname.Hash.find cfg pname) then
+    L.die InternalError "Creating two procdescs with different signatures for the same procname." ;
```

Comment:
I am not sure this will work. The backend will lookup the proc desc using the proc name only so adding more than one here is breaking the invariant expected by the backend.

However, the Java frontend also looks up the method by name following the Java class loading mechanism. So this case should not happen in theory and there may be a bug somewhere else that we would need to fix.

Also, we may want to hide the details of this in the CFG module. It should not be the responsibility of every frontend to make sure there is only one proc desc per proc name.